horovod/torch/optimizer.py

```diff
@@ -113,7 +116,18 @@ def _allreduce_grad_async(self, p):
         tensor = p.grad
         tensor_compressed, ctx = self._compression.compress(tensor)
 
-        handle = allreduce_async_(tensor_compressed, name=name, op=self.op)
+        if self.op == Average and self.gradient_predivide_factor != 1.0:
```

Comment:
Since `allreduce_async_` is in the public API, I think we should consider how this affects the user's use this function.  Here we're defining an Average op as a Sum op with a post-scale factor scaled by 1 / size.  But if the user were to call `allreduce_async_` directly with `op=Average` and `postscale_factor=s` they would get a different result.

In the implementation of `_allreduce_async`, it seems that the divisor is only size when ROCm is used.  Which would seem to break `op=Average` when calling `allreduce_async_` directly.  Maybe we can pull this logic into `_allreduce_async` to simplify things?

Reply:
This is a similar to TF, in that in this section of the code, the user does not have explicit control over the pre- and post-scale factors but only indirect control through `gradient_predivide_factor`. 

For the `allreduce_async_` implementation, they could set `op=Average` and both scale factors and it should work fine (i.e. the tensors will be prescaled before allreduce by `prescale_factor`, postscaled after allreduce by `postscale_factor` and divided by `size` for the average. When `op==Average`, we do not use divisor, but instead apply the additional `size` factor into `postscale_factor` in `EnqueueTensorAllreduce` here: https://github.com/horovod/horovod/blob/b0e51db4e26ea5be153cf3f3489660ebd0e310e3/horovod/common/operations.cc#L840

I'm can't think of a situation where someone would want to mix pre- and post-scale factors with `op==Average` but I do think this is doing what would be expected in that case.